divi/src/main.cpp

```diff
@@ -2854,6 +2868,42 @@ static bool ActivateBestChainStep(CValidationState& state, CBlockIndex* pindexMo
     return true;
 }
 
+namespace {
+
+/**
+ * Sends inv messages to update the given node about a new chain tip we have.
+ * Normally this just sends an inv for the new block; but if the node is behind
+ * (or e.g. we just had a reorg), make sure to send all inv's in order from
+ * what we know to be the node's best known block.
+ */
+void RelayNewTipToNode(CNode& node, const CBlockIndex& index)
+{
+    const auto* state = State(node.GetId());
+
+    /* In the most common case, the node is behind just one block.  Just relay
+       the current one then; also do this if we don't know where the node is.  */
+    if (state == nullptr || state->pindexBestKnownBlock == nullptr || index.pprev->GetBlockHash() == state->pindexBestKnownBlock->GetBlockHash()) {
+        node.PushInventory(CInv(MSG_BLOCK, index.GetBlockHash()));
+        return;
+    }
+
+    /* If the node is behind more than that, look for the fork point and relay
+       all inv's since then in order.  */
+    LOCK(cs_main);
+    const auto* pindexFork = chainActive.FindFork(state->pindexBestKnownBlock);
+    pindexFork = chainActive.Next(pindexFork);
+    assert(pindexFork != nullptr);
```

Comment:
In the event that a node connects to another node that has a different genesis block (cause people do this sometimes haha), pindexFork will be null, and the result of invoking .Next will be a nullpointer exception (tracing the sequence of calls) so it won't even make it to the assert before crashing.

Reply:
Good point.  I think that strictly speaking that still can't happen, because if the genesis block is not the same, then the `pindexBestKnownBlock` would not be set, as it could not be inside our `mapBlockIndex`.  But it definitely can't hurt to have an explicit check for this, which I've added in a fixup commit now.